rag/llm/chat_model.py

```
@@ -2437,16 +2454,17 @@ async def _exec_tool(tc):
-                        for tc, name, args, result, err in results:
-                            if name in _terminal and not err:
-                                logging.info(f"[Tool loop] The {name} tool produced the final answer — done.")
-                                out = result if isinstance(result, str) else json.dumps(result, ensure_ascii=False)
-                                if out:
-                                    yield out
-                                yield total_tokens
-                                return
+                    # `rag` is always terminal (dialog_service sets terminal_tools,
+                    # but a probe wrapper may drop it, so default to {"rag"}).
+                    _terminal = getattr(self, "terminal_tools", None) or {"rag"}
+                    for tc, name, args, result, err in results:
```

Ruff (0.16.2)

[warning] Loop control variable `tc` not used within loop body

(B007)